File: apps/api/src/main/kotlin/com/productivityos/goal/Goal.kt
package com.productivityos.goal

import java.time.Instant
import java.time.LocalDate
import java.util.UUID

data class Goal(
    val id: UUID,
    val userId: UUID,
    val title: String,
    val description: String? = null,
    val status: GoalStatus = GoalStatus.DRAFT,
    val deadline: LocalDate? = null,
    val completedAt: Instant? = null
) {
    fun activate(): Goal {
        require(status == GoalStatus.DRAFT) { "Only Draft goals can be activated" }
        return copy(status = GoalStatus.ACTIVE)
    }

    fun complete(now: Instant): Goal {
        require(status == GoalStatus.ACTIVE) { "Only Active goals can be completed" }
        return copy(status = GoalStatus.COMPLETED, completedAt = now)
    }

    fun reopen(): Goal {
        require(status == GoalStatus.COMPLETED) { "Only Completed goals can be reopened" }
        return copy(status = GoalStatus.ACTIVE)
    }

    fun archive(): Goal {
        require(status == GoalStatus.COMPLETED) { "Only Completed goals can be archived" }
        return copy(status = GoalStatus.ARCHIVED)
    }
}

File: apps/ios/ProductivityOS/Models/Goal.swift
import Foundation

/// Goal status from the backend `GoalStatus` enum
/// (`DRAFT`, `ACTIVE`, `COMPLETED`, `ARCHIVED`).
public enum GoalStatus: String, Codable, CaseIterable, Sendable {
    case draft = "DRAFT"
    case active = "ACTIVE"
    case completed = "COMPLETED"
    case archived = "ARCHIVED"
}

/// Goal domain model matching backend `GoalResponse`
/// (apps/api `goal/dto/GoalResponse.kt`). The backend does not expose a
/// `progress` field — progress is derived client-side from the goal's
/// projects and their tasks.
public struct Goal: Identifiable, Codable, Hashable, Sendable {
    public let id: UUID
    public let userId: UUID
    public let title: String
    public let description: String?
    public let status: GoalStatus
    public let deadline: Date?
    public let completedAt: Date?
    public let createdAt: Date
    public let updatedAt: Date

    public init(
        id: UUID,
        userId: UUID,
        title: String,
        description: String? = nil,
        status: GoalStatus = .active,
        deadline: Date? = nil,
        completedAt: Date? = nil,
        createdAt: Date,
        updatedAt: Date
    ) {
        self.id = id
        self.userId = userId
        self.title = title
        self.description = description
        self.status = status
        self.deadline = deadline
        self.completedAt = completedAt
        self.createdAt = createdAt
        self.updatedAt = updatedAt
    }
}